hydra-node/src/Hydra/Chain/Direct/Handlers.hs

```diff
@@ -145,11 +162,27 @@ mkChain tracer queryTimeHandle wallet ctx LocalChainState{getLatest} submitTx =
         submitTx vtx
     , -- Handle that creates a draft commit tx using the user utxo.
       -- Possible errors are handled at the api server level.
-      draftTx = \utxo -> do
+      draftCommitTx = \utxoInputs -> do
         chainState <- atomically getLatest
         case Hydra.Chain.Direct.State.chainState chainState of
-          Initial st ->
-            sequenceA $ finalizeTx wallet ctx chainState utxo <$> commit ctx st utxo
+          Initial st -> do
+            walletUtxos <- atomically getUTxO
+            let walletTxIns = fromLedgerTxIn <$> Map.keys walletUtxos
+            let userTxIns = Set.toList $ UTxO.inputSet regularUTxO
+            let matchedWalletUtxo = filter (`elem` walletTxIns) userTxIns
+            if null matchedWalletUtxo
+              then sequenceA $ finalizeTx wallet ctx chainState (regularUTxO <> scriptUTxO) <$> buildCommitTx
+              else pure $ Left SpendingNodeUtxoForbidden
+           where
+            (regularUTxO, scriptUTxO, scriptWitnesses) = prepareCommitTxInputs utxoInputs
+            buildCommitTx = do
+              txBody <- commitTxBody ctx st regularUTxO
+              pure . unsafeBuildTransaction $
+                txBody
+                  & addInputs
+                    ( second (BuildTxWith . ScriptWitness ScriptWitnessForSpending)
+                        <$> scriptWitnesses
+                    )
```

Comment:
This should stay very simple code without much additional logic (maybe except the checking for wallet utxo).

To achieve that, do not build the transaction half-way and extend it here, but pass in a more complete type right away. That is, not only `UTxO` which is "regular utxo" as you call it, but using `UTxO' (TxOut CtxUTxO, Maybe ScriptWitness)` (or something along those lines - a conversion from `ScriptInfo` to the cardano-api ScriptWitness would make sense maybe on the API layer already).

Then, in the `Direct.Tx` module, we change the signature of `commitTx` to take this `UTxO' (TxOut CtxUTxO, Maybe ScriptWitness)` (we could think about a type alias `type UTxOToCommit = UTxO' (TxOut CtxUTxO, Maybe ScriptWitness)`) and convert that accordingly to `TxIns` to construct the right transaction body.

Reply:
We are not longer building the transaction half-way, instead we are passing a more complex type as requested and made the mapping from ScriptInfo -> ScriptWitness also in the API layer.

Still missing changes in `Direct.Tx` module tho.